### ISSUE TITLE
Clear Instructions for Finetuning Faces

### ISSUE THREAD

**Issue by lazy-nurd** [adobe-research/custom-diffusion#13]
Hey thanks for amazing work,
I was trying to implement finetuning faces like dreambooth works really well in finetuning model on specific face subject. But when I tried finetuning faces there is very less guidelines for it. But I tried training through diffusers and the results were very odd and the identity loss was very high. Training through scripts is difficult to understand since there is no specific guideline for what is instance prompt and what is class prompt. 
It would be great if you can provide us better guidelines on finetuning faces easily with custom diffusion. 
Also the papers say the model is ~6x faster than dreambooth, however this does not seems to be the case, since training custom diffusion through diffusers is taking almost the same time and gpu memory, please guide us on this as well.
Really looking forward to your reply

**Comment by nupurkmr9** [comment]
Hi,
Thanks for your interest in our work. I have listed below the answers to your questions. Let me know if you have any other questions.  

**class_prompt**: the category of the training images, e.g., `person` for human faces. 
**instance_prompt**: the text prompt used during training on the target training images, i.e., `photo of a <new1> person` 
**training time**: all our experiments in the paper were with stable-diffusion-1.4 and 2A100 GPUs, on which it takes ~6 minutes and 30GB VRAM. For human faces, a lower learning rate and more training steps are better, which can then take ~18 minutes of training. This can be reduced by using less batch size but more training time, or using `deepspeed` library with `accelarate`. Can you mention which GPU configuration you are training the model on?
**regularization images**: we collect real images from the LAION dataset using [clip-retrieval](https://github.com/rom1504/clip-retrieval) with similar captions as class_prompt before starting the training. This can take some time, depending on the class prompt. 
**Number of images for human faces**: I haven't benchmarked the performance on human faces extensively. But in my experiments, 15-20 high-resolution images should be good, more is better. An example result is shown [here](https://github.com/adobe-research/custom-diffusion#fine-tuning-on-human-faces).  

For training, using these configurations should be good. 
```
accelerate launch src/diffuser_training.py \
          --pretrained_model_name_or_path="CompVis/stable-diffusion-v1-4"  \
          --instance_data_dir=./data/cat  \
          --class_data_dir=./real_reg/samples_cat/ \
          --output_dir=./logs/cat  \
          --with_prior_preservation --real_prior --prior_loss_weight=1.0 \
          --instance_prompt="photo of a <new1> person"  \
          --class_prompt="person" \
          --resolution=512  \
          --train_batch_size=2  \
          --learning_rate=5e-6  \
          --lr_warmup_steps=0 \
          --max_train_steps=750 \
          --num_class_images=200 \
          --scale_lr \
          --modifier_token "<new1>"
 ```

**Comment by nupurkmr9** [comment]
Hi,
Thanks for pointing this out. Can you try increasing the aesthetic weight in this [line](https://github.com/adobe-research/custom-diffusion/blob/5cd1e9c869b793d88573533a6a2adccd10aadcd0/src/retrieve.py#L15)? 
I will update the readme/code to allow that with a flag easily. 

Also, our method works with generated images as regularization as well which can be enabled by removing the `--real_prior` flag from the command.

**Comment by SaimAli420** [comment]
> Hi, Thanks for your interest in our work. I have listed below the answers to your questions. Let me know if you have any other questions.
> 
> **class_prompt**: the category of the training images, e.g., `person` for human faces. **instance_prompt**: the text prompt used during training on the target training images, i.e., `photo of a <new1> person` **training time**: all our experiments in the paper were with stable-diffusion-1.4 and 2A100 GPUs, on which it takes ~6 minutes and 30GB VRAM. For human faces, a lower learning rate and more training steps are better, which can then take ~18 minutes of training. VRAM requirement can be reduced by using less batch size but more training time, or using `deepspeed` library with `accelarate`. Can you mention which GPU configuration you are training the model on? **regularization images**: we collect real images from the LAION dataset using [clip-retrieval](https://github.com/rom1504/clip-retrieval) with similar captions as class_prompt before starting the training. This can take some time, depending on the class prompt. **Number of images for human faces**: I haven't benchmarked the performance on human faces extensively. But in my experiments, 15-20 high-resolution images should be good, more is better. An example result is shown [here](https://github.com/adobe-research/custom-diffusion#fine-tuning-on-human-faces).
> 
> For training, using these configurations should be good.
> 
> ```
> accelerate launch src/diffuser_training.py \
>           --pretrained_model_name_or_path="CompVis/stable-diffusion-v1-4"  \
>           --instance_data_dir=./data/cat  \
>           --class_data_dir=./real_reg/samples_cat/ \
>           --output_dir=./logs/cat  \
>           --with_prior_preservation --real_prior --prior_loss_weight=1.0 \
>           --instance_prompt="photo of a <new1> person"  \
>           --class_prompt="person" \
>           --resolution=512  \
>           --train_batch_size=2  \
>           --learning_rate=5e-6  \
>           --lr_warmup_steps=0 \
>           --max_train_steps=750 \
>           --num_class_images=200 \
>           --scale_lr \
>           --modifier_token "<new1>"
> ```

Hi,

I appreciate your work. I used same command as you mentioned here.
Model takes at least 3000 steps to  recover the identity loss.
Not produce good results on too long prompts.
Finetune with "runwayml/stable-diffusion-v1-5" producing good results than 2.1 version but not too good.

Some of the results comparison between 1.5 vs 2.1 versions.
**1.5 prompts:**
Ultra HD quality of styriodio woman as god of war character with beautiful open eyes 
beautiful styriodio woman as viking.

**Best Results:**
![1673271198 6342075](https://user-images.githubusercontent.com/73534805/211345780-abc51afa-e189-4178-bd9a-6015c58535b1.png)
![1673271521 3364086](https://user-images.githubusercontent.com/73534805/211345804-86234475-fefe-4e76-8c6a-c7f332f04890.png)
![1673271766 6419137](https://user-images.githubusercontent.com/73534805/211345815-5af71101-3dcf-48ac-b568-7bcb718db43f.png)
![1673272051 3470387](https://user-images.githubusercontent.com/73534805/211345820-5f509f6c-08b4-4a1a-b8e7-13a08c124c97.png)


**2.1 prompts:**
Ultra HD quality of styriodio woman as god of war character with beautiful open eyes 
beautiful styriodio woman as viking. 

**Best Results:**
![1673274295 3310952](https://user-images.githubusercontent.com/73534805/211348601-373d37fb-918e-4c10-abff-e89e143fb5b1.png)
![1673274302 77981](https://user-images.githubusercontent.com/73534805/211348647-a7433e76-99bb-4ba1-827a-9c40b0afaa59.png)
![1673274549 0818434](https://user-images.githubusercontent.com/73534805/211348659-fc1a364a-efd5-4e3a-93be-0dbda59af0e5.png)
![1673275145 9398391](https://user-images.githubusercontent.com/73534805/211348665-bab2b6f1-63ca-456e-8a6c-747be8e44050.png)

**On long prompts:**
Portrait painting of styriodio as a pretty goth girl, gorgeous tattoos, beautiful hair, symmetrical! intricate, elegant, highly detailed, digital painting, artstation, concept art, smooth, sharp focus, illustration, art by artgerm and greg rutkowski and alphonse mucha
![1673279500 9328053](https://user-images.githubusercontent.com/73534805/211351303-eba440bc-c72d-4f85-b153-4ef865621ce2.png)
![1673279508 5792706](https://user-images.githubusercontent.com/73534805/211351354-e5f7d644-d1c9-4639-b636-0cd5d5e0131b.png)
![1673279516 418743](https://user-images.githubusercontent.com/73534805/211351368-a50a33d8-51c6-4631-bd77-1cafea9d04fc.png)
![1673279524 236609](https://user-images.githubusercontent.com/73534805/211351374-27f578d2-eb29-4222-b511-37e43dda2acd.png)
![1673279532 0180798](https://user-images.githubusercontent.com/73534805/211351382-a57be9d3-8329-48e9-bbd6-74f0fe379229.png)

Total identity loss on Emma Watson Images. But dreambooth working fine on these long prompts.

Can you please guide me how I improve the model on long prompts.
Also guide me about how I improve the results. Because after 3000 steps still there are some identity loss.
Looking forward to your reply thanks.

**Comment by dawei03896** [comment]
@nupurkmr9  Thanks for amazing wonderful work, I used the these configurations, but the result is bad. 

accelerate launch src/diffuser_training.py \
          --pretrained_model_name_or_path=$MODEL_NAME  \
          --instance_data_dir=$INSTANCE_DIR  \
          --class_data_dir=$CLASS_DIR \
          --output_dir=$OUTPUT_DIR  \
          --with_prior_preservation --real_prior --prior_loss_weight=1.0 \
          --instance_prompt="photo of a <new1> person"  \
          --class_prompt="person" \
          --resolution=512  \
          --train_batch_size=2  \
          --learning_rate=5e-6  \
          --lr_warmup_steps=0 \
          --max_train_steps=3000 \
          --num_class_images=200 \
          --scale_lr --hflip  \
          --mixed_precision "no" \
          --modifier_token "<new1>"

**Comment by nupurkmr9** [comment]
Hi, 

I'm sorry for the delayed response in this thread. I trained a model on 16 face photos with the following configurations on  2 GPUs. Training on faces with the `--freeze_model crossattn` configuration is usually better than the default setting. 

```
export MODEL_NAME="CompVis/stable-diffusion-v1-4"
CUDA_VISIBLE_DEVICES=1,2 accelerate launch src/diffuser_training.py \
          --pretrained_model_name_or_path=$MODEL_NAME  \
          --instance_data_dir=./face_photos  \
          --class_data_dir=./real_reg/samples_person  \
          --output_dir=./face_model  \
          --with_prior_preservation --real_prior --prior_loss_weight=1.0 \
          --instance_prompt="photo of a <new1> person"  \
          --class_prompt="person" \
          --resolution=512  \
          --train_batch_size=2  \
          --learning_rate=5e-6  \
          --lr_warmup_steps=0 \
          --max_train_steps=2000 \
          --num_class_images=200 \
          --scale_lr  --hflip \
          --freeze_model crossattn \
          --modifier_token "\<new1\>"
```

And these are the results with sampling done as `python src/sample_diffuser.py --prompt  <text-prompt> --delta_ckpt <path-to/delta.bin> --ckpt "CompVis/stable-diffusion-v1-4" --freeze_model crossattn --batch_size 10`. I selected the best 3 out of 10 generations. I used two prompts from the above post. 

![face_results](https://user-images.githubusercontent.com/9297728/219546004-cb2994d8-0a8b-4764-9170-653283faec93.png)
